### PR TITLE
Integrate Mayhem

### PR DESCRIPTION
**Pull request by rnshah9**

### CODE REVIEW

**Review by xansec — CHANGES_REQUESTED** [review]
Hmm, this target doesn't quite look correct. I notice that it's crashing on every input: https://mayhem.forallsecure.com/rnshah9/mayhem-micropython/micropython/5. I don't think just sending fuzzed input to the python interpreter will work here. Is there a module that specifically parses input that we could fuzz instead? I'm thinking of something from here: http://docs.micropython.org/en/latest/reference/repl.html Please make the requested changes.

**Comment by rnshah9** [comment]
This project has a pretty complex build, so after spending a good amount time trying to build a fuzzer executable manually I decided to just replace the `main` function with the harness and reuse the project's build.

But `main.c` has a good amount of functionality that breaks the build if excluded, so I just copied the entire file (excl. `main`) into `fuzz_qstr_compute_hash.c`.

File: mayhem/Mayhemfile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+project: micropython
```
Comment (xansec) [inline mayhem/Mayhemfile:1]: We'll overwrite the image name here.

File: mayhem/fuzz_qstr_compute_hash.c [hunk 2]
```diff
@@ -0,0 +1,878 @@
+#include <stdint.h>
```
Comment (xansec) [inline mayhem/fuzz_qstr_compute_hash.c:1]: Hmm, I don't necessarily love this approach, as we now have static copies of every function. If the `main.c` changes at all, how will those changes be propogated to this harness? Is there truly no way to run just the fuzzing functionality apart from the rest of the code (i.e. including/building libraries in Cmake/make style library dependencies?)

**Review by xansec — CHANGES_REQUESTED** [review]
Please make the requested changes.

File: mayhem/fuzz_qstr_compute_hash.c (hunk 2 above)
Comment (rnshah9) [reply mayhem/fuzz_qstr_compute_hash.c:1]: I agree it's suboptimal. One other option would be to edit `main.c` directly in this fork and let git handle the merging of upstream changes to the file.

Could you elaborate on "including/building libraries..."? Do you mean trying to build this project as a library?

**Comment by xansec** [comment]
Resolved with https://github.com/mayhemheroes/micropython/pull/2

**Pull request closed without merging**